python/lsst/source/injection/inject_engine.py

```diff
@@ -0,0 +1,462 @@
+        Prefix to add to the stamp name.
+    logger : `lsst.utils.logging.LsstLogAdapter`, optional
+        Logger to use for logging messages.
+
+    Yields
+    ------
+    sky_coords : `lsst.geom.SpherePoint`
+        RA/Dec coordinates of the source.
+    pixel_coords : `lsst.geom.Point2D`
+        Pixel coordinates of the source.
+    draw_size : `int`
+        Size of the stamp to draw.
+    object : `galsim.transform.Transformation`
+        A fully specified and transformed GalSim object.
+    """
+    if logger:
+        source_types = Counter(injection_catalog["source_type"])  # type: ignore
+        grammar0 = "source" if len(injection_catalog) == 1 else "sources"
+        grammar1 = "type" if len(source_types) == 1 else "types"
+        logger.info(
+            "Generating %d injection %s consisting of %d unique %s: %s.",
+            len(injection_catalog),
+            grammar0,
+            len(source_types),
+            grammar1,
+            ", ".join(f"{k}({v})" for k, v in source_types.items()),
+        )
+    for source_data_full in injection_catalog:
+        items = dict(source_data_full).items()  # type: ignore
+        source_data = {k: v for (k, v) in items if v is not ma.masked}
+        try:
+            sky_coords = SpherePoint(float(source_data["ra"]), float(source_data["dec"]), degrees)
+        except KeyError:
+            sky_coords = wcs.pixelToSky(float(source_data["x"]), float(source_data["y"]))
+        try:
+            pixel_coords = Point2D(source_data["x"], source_data["y"])
+        except KeyError:
+            pixel_coords = wcs.skyToPixel(sky_coords)
+        try:
+            inst_flux = photo_calib.magnitudeToInstFlux(source_data["mag"], pixel_coords)
+        except LogicError:
+            continue
+        try:
+            draw_size = int(source_data["draw_size"])
+        except KeyError:
+            draw_size = 0
+
+        if source_data["source_type"] == "Stamp":
+            stamp_file = stamp_prefix + source_data["stamp"]
+            object = make_galsim_stamp(stamp_file, fits_alignment, wcs, sky_coords, inst_flux)
+        elif source_data["source_type"] == "Trail":
+            object = make_galsim_trail(source_data, wcs, sky_coords, inst_flux)
+        else:
+            object = make_galsim_object(source_data, source_data["source_type"], inst_flux)
+
+        yield sky_coords, pixel_coords, draw_size, object
+
+
+def make_galsim_object(
+    source_data: dict[str, Any],
+    source_type: str,
+    inst_flux: float,
+) -> galsim.transform.Transformation:
+    """Make a generic GalSim object from a collection of source data.
+
+    Parameters
+    ----------
+    source_data : `dict` [`str`, `Any`]
+        Dictionary of source data.
+    source_type : `str`
+        Type of the source, corresponding to a GalSim class.
+    inst_flux : `float`
+        Instrumental flux of the source.
+
+    Returns
+    -------
+    object : `galsim.transform.Transformation`
```

Comment:
I think `Transformation` here is an implementation detail.  I'd just say this is a `galsim.gsobject.GSObject`.

Reply:
Modified here and elsewhere in this file.